server/src/services/assistantService.js

```diff
@@ -0,0 +1,185 @@
+    throw error;
+  }
+}
+
+// Add message to thread
+async function addToThread(threadID, message) {
+  try {
+    await openai.beta.threads.messages.create(threadID, {
+      role: "user",
+      content: message,
+    });
+  } catch (error) {
+    console.error("Error adding to thread:", error);
+    throw error;
+  }
+}
+
+// Create a run
+async function createRun(threadID, assistantID) {
+  try {
+    let run = await openai.beta.threads.runs.createAndPoll(threadID, {
+      assistant_id: assistantID,
+    });
+    return run;
+  } catch (error) {
+    console.error("Error creating run: ", error);
+    throw error;
+  }
+}
+
+// Handles required_action run status
+async function handleRequiredAction(run, thread) {
+  if (
+    run.required_action &&
+    run.required_action.submit_tool_outputs &&
+    run.required_action.submit_tool_outputs.tool_calls
+  ) {
+    const toolOutputs = await Promise.all(
+      run.required_action.submit_tool_outputs.tool_calls.map(async (tool) => {
+        if (tool.function.name === "get_weather") {
+          const location = tool.function.arguments;
+          console.log("Location requested:", location);
+          const weatherInfo = await get_weather(location);
+          console.log("Weather info:", weatherInfo);
+          return {
+            tool_call_id: tool.id,
+            output: weatherInfo,
+          };
+        }
+      })
+    );
+
+    console.log("Tool outputs:", toolOutputs);
+
+    if (toolOutputs.length > 0) {
+      run = await openai.beta.threads.runs.submitToolOutputsAndPoll(
+        thread.id,
+        run.id,
+        { tool_outputs: toolOutputs },
+      );
+      console.log("Tool outputs submitted successfully.");
+    } else {
+      console.log("No tool outputs to submit.");
+    }
+
+    return handleRunStatus(run, thread);
+  }
+}
+
+// Handle run status
+const handleRunStatus = async (run, thread) => {
+  // Check if the run is completed
+  if (run.status === "completed") {
+    let messages = await openai.beta.threads.messages.list(thread.id);
+    console.log(messages.data);
+    return messages.data;
+  } else if (run.status === "requires_action") {
+    console.log(run.status);
+    return await handleRequiredAction(run, thread);
+  } else {
+    console.error("Run did not complete:", run);
+  }
+};
+
+// Retrieve and display messages
+async function displayMessages(thread) {
+  try {
+    const threadMessages = await openai.beta.threads.messages.list(thread.id);
+
+    threadMessages.data.forEach(message => {
+      if (message.role === 'user') {
+        console.log(`User: ${message.content[0].text.value}`);
+      } else if (message.role === 'assistant') {
+        console.log(`Assistant: ${message.content[0].text.value}`);
+      } else if (message.role === 'function') {
+        console.log(`Function: ${message.content[0].text.value}`);
+      }
+    });
+
+  } catch (error) {
+    console.error("Error displaying messages:", error);
+    throw error;
+  }
+}
```

Comment:
What is the purpose of this function, can't this be tested with swagger and postman?

Reply:
This simply log the messages of user and assistant. I didn't understand how to have the Swagger thing add said messages as a response in Swagger, and I wanted a quick way to test if everything was working as intended.

Reply:
> This simply log the messages of user and assistant. I didn't understand how to have the Swagger thing add said messages as a response in Swagger, and I wanted a quick way to test if everything was working as intended.

Swagger simply logs the output, the `res` can be anything you define.

Reply:
I don't entirely understand, but I'll try to have Swagger take user-defined message and outputs.